File: server/work_dating.py
"""Source-calendar dates. Index keys order written dates, never converted instants."""
import re


def partial_date(value, calendar=None):
    if not isinstance(value, str) or not re.fullmatch(r'\d{3,4}(?:-\d{2}(?:-\d{2})?)?', value):
        return None
    parts = [int(v) for v in value.split('-')]
    y, m, d = (parts + [None, None])[:3]
    if y < 1 or (m is not None and not 1 <= m <= 12):
        return None
    leap = not (calendar == 'swedish' and y == 1700) and y % 4 == 0 and (calendar != 'gregorian' or y % 100 != 0 or y % 400 == 0)
    days = (30 if calendar in (None, 'swedish') and y == 1712 else 29 if leap else 28) if m == 2 else 30 if m in (4, 6, 9, 11) else 31
    if d is not None and not 1 <= d <= days:
        return None
    return parts


def bound(value, upper=False):
    parts = partial_date(value)
    if not parts:
        return None
    y, m, d = (parts + [None, None])[:3]
    return y * 10000 + (m or (12 if upper else 1)) * 100 + (d or (31 if upper else 1))


def parse_dating_text(raw):
    s = str(raw or '').strip()
    def endpoint(v):
        if partial_date(v):
            return v
        match = re.fullmatch(r'(\d{1,2})[.-](\d{1,2})[.-](\d{4})', v)
        if match:
            d, m, y = match.groups()
            iso = f'{y}-{int(m):02}-{int(d):02}'
            if partial_date(iso):
                return iso
        return None
    one = endpoint(s)
    if one:
        return {'start': one}
    pair = re.split(r'\s+[–—-]\s+|[–—/]', s)
    if len(pair) != 2:
        match = re.fullmatch(r'(\d{3,4})-(\d{3,4})', s)
        pair = list(match.groups()) if match else []
    if len(pair) == 2:
        start, end = [endpoint(v.strip()) for v in pair]
        if start and end and bound(start) <= bound(end, True):
            return {'start': start, 'end': end}
    return None


def clean_dating(value):
    if value is None:
        return None
    if not isinstance(value, dict):
        raise ValueError('Invalid dating')
    if 'end' in value and not value['end']:
        raise ValueError('Invalid dating end')
    allowed = {'start', 'end', 'kind', 'calendar', 'approximate', 'note', 'source_text'}
    result = {k: v for k, v in value.items() if k in allowed and v not in (None, '')}
    calendar = result.get('calendar')
    if calendar not in (None, 'julian', 'gregorian', 'swedish'):
        raise ValueError('Invalid dating calendar')
    if result.get('kind') not in (None, 'span', 'uncertain'):
        raise ValueError('Invalid dating kind')
    if 'approximate' in result and not isinstance(result['approximate'], bool):
        raise ValueError('Invalid dating approximation')
    if any(not isinstance(result[k], str) for k in ('note', 'source_text') if k in result):
        raise ValueError('Invalid dating text')
    if not partial_date(result.get('start'), calendar):
        raise ValueError('Invalid dating start')
    end = result.get('end')
    if end is not None and (not partial_date(end, calendar) or bound(result['start']) > bound(end, True)):
        raise ValueError('Invalid dating end')
    return result


def dating_updates(updates):
    """Explicit dating edits own legacy compatibility fields; omission preserves them."""
    if 'dating' not in updates:
        return updates
    result = dict(updates)
    value = clean_dating(result['dating'])
    result['dating'] = value
    if value:
        start, end = value['start'], value.get('end')
        result['year'] = (int(start.split('-')[0]) + int((end or start).split('-')[0])) // 2
        result['year_display'] = start + (' / ' + end if end else '')
    return result


def index_dating(meta, year_range):
    value = meta.get('dating')
    if value:
        value = clean_dating(value)
    else:
        value = parse_dating_text(meta.get('year_display'))
        if value:
            value['source_text'] = meta['year_display']
    if value:
        start = bound(value['start'])
        end = bound(value.get('end') or value['start'], True)
    elif year_range:
        start, end = year_range[0] * 10000 + 101, year_range[1] * 10000 + 1231
    else:
        start = end = 0
    return {'dating': value, 'date_start': start, 'date_end': end, 'date_sort': start}

File: src/utils/workDating.ts
/** Dates as written on the source: no timezone or implicit calendar conversion. */
export interface WorkDating {
  start: string;
  end?: string;
  kind?: 'span' | 'uncertain';
  calendar?: 'julian' | 'gregorian' | 'swedish';
  approximate?: boolean;
  note?: string;
  source_text?: string;
}

export function parsePartialDate(value: string, calendar?: string): number[] | null {
  if (!/^\d{3,4}(?:-\d{2}(?:-\d{2})?)?$/.test(value)) return null;
  const parts = value.split('-').map(Number);
  const [y, m, d] = parts;
  if (y < 1 || (m !== undefined && (m < 1 || m > 12))) return null;
  // Unknown calendars allow Julian leap days; Sweden had 30 February 1712.
  const leap = !(calendar === 'swedish' && y === 1700) && y % 4 === 0 && (calendar !== 'gregorian' || y % 100 !== 0 || y % 400 === 0);
  const days = m === 2 ? ((!calendar || calendar === 'swedish') && y === 1712 ? 30 : leap ? 29 : 28) : [4, 6, 9, 11].includes(m) ? 30 : 31;
  if (d !== undefined && (d < 1 || d > days)) return null;
  return parts;
}

export function dateBound(value: string | number | undefined, upper = false): number | null {
  if (value === undefined || value === '') return null;
  const parts = parsePartialDate(String(value));
  if (!parts) return null;
  const [y, m, d] = parts;
  // 31 is an inclusive search sentinel, not an asserted date.
  return y * 10000 + (m ?? (upper ? 12 : 1)) * 100 + (d ?? (upper ? 31 : 1));
}

/** Only unambiguous numeric legacy dates are inferred; prose remains untouched. */
export function parseDatingText(raw: string): WorkDating | null {
  const s = raw.trim();
  const endpoint = (v: string): string | null => {
    if (parsePartialDate(v)) return v;
    const m = v.match(/^(\d{1,2})[.-](\d{1,2})[.-](\d{4})$/);
    const iso = m ? `${m[3]}-${m[2].padStart(2, '0')}-${m[1].padStart(2, '0')}` : '';
    return parsePartialDate(iso) ? iso : null;
  };
  const one = endpoint(s);
  if (one) return { start: one };
  const pair = s.split(/\s+[–—-]\s+|[–—/]/);
  if (pair.length === 2) {
    const start = endpoint(pair[0].trim()), end = endpoint(pair[1].trim());
    if (start && end && dateBound(start)! <= dateBound(end, true)!) return { start, end };
  }
  const years = s.match(/^(\d{3,4})-(\d{3,4})$/);
  if (years && Number(years[1]) <= Number(years[2])) return { start: years[1], end: years[2] };
  return null;
}

export function datingText(dating: WorkDating): string {
  return dating.start + (dating.end ? ` / ${dating.end}` : '');
}

export function datingError(dating?: WorkDating | null): boolean {
  return !!dating && (!parsePartialDate(dating.start, dating.calendar) ||
    (dating.end !== undefined && (!parsePartialDate(dating.end, dating.calendar) || dateBound(dating.start)! > dateBound(dating.end, true)!)));
}

export function hasImpreciseDating(dating: WorkDating | null | undefined, display?: string | null): boolean {
  const d = dating ?? parseDatingText(display || '');
  return !d || !!d.approximate || d.kind === 'uncertain' || d.start.split('-').length < 3 || (!!d.end && d.end.split('-').length < 3);
}
